predicate/solver/ast.py

```diff
@@ -675,6 +713,63 @@ def traverse(self):
         ) == z3.BoolVal(True)
 
 
+class StringListForEach(LogicMixin):
+    def __init__(self, string_list: StringList, for_each):
+        self.string_list = string_list
+        self.for_each = for_each
+
+    def walk(self, fn):
+        fn(self)
+        self.for_each(StringListElement(self.string_list)).walk(fn)
+
+    def __str__(self):
+        return """({}.for_each({}))""".format(
+            self.string_list, self.for_each(StringListElement(self.string_list))
+        )
+
+    def traverse(self):
+        uid = "string_list_for_each_{}".format(id(self))
+        fn_string_list_for_each = z3.RecFunction(uid, StringListSort, z3.BoolSort())
+
+        vals = z3.Const(uid + "_vals", StringListSort)
+
+        # We have generated an expression from lambda,
+        #
+        # and passed this expression into the recursive function definition
+        #
+        # Given the lambda function
+        #
+        # for_each(x: x == "potato"),
+        #
+        # We would get:
+        #
+        # StringListSort.car(vals) == z3.StringVal("potato")
+        #
+        # We pass this expression into recursive function definition below,
+        # and we call this anonymous recursive function on our list parameter.
+
+        z3.RecAddDefinition(
```

Comment:
should we declare this outside of this function globally to avoid double declarations which could cause wierd issues, hard to read models etc? We seem to do this in other places so why not follow the pattern?

Reply:
If you spot it above, this creates unique functions for each call, because each `for_each` call with lambda requires a separate recursive function body. So this case is special, the function name is also unique if you see the name above it uses `id` function that returns unique object id

Reply:
Quick update - I replaced usage of `id` function with proper `uuid`, as I'm not sure if we could get any collisions with ID

Reply:
👍